### PR TITLE
fix: authenticate Redis connections with user and password (#3309)

### LINKED ISSUE #3309: 🐛 Bug Report: NOAUTH Authentication RedisException thrown
I've been working on adding possibility of Procfile based deployments so one could deploy to Heroku or Dokku keeping in mind the entrypoints for all required containers e.g. worker-deletes and so fourth.

Ultimately, the Procfile should look like following

```
web: php -e app/http.php -dopcache.preload=opcache.preload=/usr/src/code/app/preload.php
worker-functions: . worker-functions
worker-audits: ...
worker-deletes: ...
```

Before I jumped into spinning those containers up, I thought to **run deployment through Dockerfile** on my Dokku server.

Initially I was getting the following error:

```
ERROR   php_swoole_server_rshutdown() (ERRNO 503): Fatal error: Uncaught Exception: Failed to connect to database: SQLSTATE[HY000] [2002] Connection refused in /usr/src/code/app/http.php:75
```

After fiddling around, I found out that the error shown is wrong. **Redis credentials had issues, not database**. I happen to proceed connecting database successfully.

The next roadblock is the following.

```
       Fatal error: Uncaught RedisException: NOAUTH Authentication required. in /usr/src/code/vendor/utopia-php/cache/src/Cache/Adapter/Redis.php:32
       Stack trace:
       #0 /usr/src/code/vendor/utopia-php/cache/src/Cache/Adapter/Redis.php(32): Redis->get('cache-_console:...')
       #1 /usr/src/code/vendor/utopia-php/cache/src/Cache/Cache.php(47): Utopia\Cache\Adapter\Redis->load('cache-_console:...', 86400)
       #2 /usr/src/code/vendor/utopia-php/database/src/Database/Database.php(761): Utopia\Cache\Cache->load('cache-_console:...', 86400)
       #3 /usr/src/code/vendor/utopia-php/database/src/Database/Database.php(379): Utopia\Database\Database->getDocument(Object(Utopia\Database\Document), 'audit')
       #4 /usr/src/code/app/http.php(104): Utopia\Database\Database->getCollection('audit')
       #5 {main}
         thrown in /usr/src/code/vendor/utopia-php/cache/src/Cache/Adapter/Redis.php on line 32
       [2022-05-30 08:43:03 #8.4]       ERROR   php_swoole_server_rshutdown() (ERRNO 503): Fatal error: Uncaught RedisException: NOAUTH Authentication required. in /usr/src/code/vendor/utopia-php/cache/src/Cache/Adapter/Redis.php:32
       Stack trace:
       #0 /usr/src/code/vendor/utopia-php/cache/src/Cache/Adapter/Redis.php(32): Redis->get('cache-_console:...')
       #1 /usr/src/code/vendor/utopia-php/cache/src/Cache/Cache.php(47): Utopia\Cache\Adapter\Redis->load('cache-_console:...', 86400)
       #2 /usr/src/code/vendor/utopia-php/database/src/Database/Database.php(761): Utopia\Cache\Cache->load('cache-_console:...', 86400)
       #3 /usr/src/code/vendor/utopia-php/database/src/Database/Database.php(379): Utopia\Database\Database->getDocument(Object(Utopia\Database\Document), 'audit')
       #4 /usr/src/code/app/http.php(104): Utopia\Database\Database->getCollection('audit')
       #5 {main}
         thrown in /usr/src/code/vendor/utopia-php/cache/src/Cache/Adapter/Redis.php on line 32
```

Could someone please help?

Thanks

### 👟 Reproduction steps
Spin up a Dokku server, assign environment variables or push the application to Heroku with environment variables all setup. 

#### Dokku
```git push dokku master:master```

#### Heroku
```git push heroku master:master```

### 👍 Expected behavior

It should connect to Redis as all the configurations are correct. I've tested it through Medis and TablePlus

### 👎 Actual Behavior

The app crashes upon deployment after the Dockerfile build is succeeded and app is about to release. 

### 🎲 Appwrite version

Version 0.14.x

### 💻 Operating system

Linux

### 🧱 Your Environment

N.A

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/HEAD/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by RushikeshGandhmal**
## What does this PR do?

Fixes #3309 — self-hosted Appwrite installations using Redis authentication (`requirepass` or ACL users) currently fail with `NOAUTH Authentication required` because several Redis connection sites ignore `_APP_REDIS_USER` and only call `auth()` when a password is set.

This PR introduces a small `Appwrite\Redis\Auth` helper that normalizes credentials (default user + password, ACL username + password, named `nopass` users, and the edge case where the username or password is literally the string `"0"`). It then wires that helper into all Redis construction sites:

- `app/init/resources.php` — project-level Redis cache connection
- `app/init/registers.php` — pools and queue broker connection
- `app/realtime.php` — realtime Redis connection
- `app/worker.php` — worker receive and command Redis queues
- `src/Appwrite\Queue\Connection\Redis.php` — temporary wrapper around the upstream Utopia Queue Redis connection that authenticates after connect (the upstream class stores the credentials but never calls `auth()`)

It also fixes `Appwrite\URL\URL::unparse()` so that DSNs generated for Redis preserve the string `"0"` as a username/password, which is required to correctly build fallback `redis://` URLs.

## Test Plan

- Unit tests added in `tests/unit/Redis/AuthTest.php` cover credential normalization and the `authenticate()` dispatcher for all common auth shapes, including `"0"` values.
- Integration test added in `tests/unit/Redis/ConnectionTest.php` creates a temporary ACL user and a temporary `nopass` ACL user on the running Redis container, then verifies the Appwrite queue wrapper and Utopia cache adapter can read/write after authentication.
- Live Redis edge-case smoke test was run against a temporary Redis container with `requirepass 0`, ACL user `0`/`0`, and a named `nopass` user; queue and cache operations succeeded.
- Full Pint run (`vendor/bin/pint --test --config pint.json`) passes: 1641 files, 0 issues.
- Targeted PHPStan analysis on all touched files passes with no errors.
- Focused PHPUnit run (`tests/unit/Redis`, `tests/unit/URL/URLTest.php`) passes: 8 tests, 24 assertions.

## Related PRs and Issues

- Fixes #3309

## Checklist

- [x] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [ ] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs? (No API metadata changes.)

## Note for the reviewer

The wrapper `src/Appwrite\Queue\Connection\Redis` is intentionally a temporary, minimal Appwrite-side shim. It can be removed once the upstream `utopia-php/queue` `Connection\Redis` class authenticates with the user/password it already stores in its constructor. The change set is otherwise limited to replacing inline `if ($pass) { $redis->auth(...); }` blocks with the new `RedisAuth::authenticate()` helper and reading `_APP_REDIS_USER` where it was previously ignored.

### CODE REVIEW

File: src/Appwrite/Redis/Auth.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    public static function authenticate(\Redis $redis, ?string $user, ?string $password): void
+    {
+        $credentials = self::credentials($user, $password);
+
+        if ($credentials === null) {
+            return;
+        }
+
+        $redis->auth($credentials);
+    }
```
[the remark this answers is not in the document]
Comment (RushikeshGandhmal) [reply src/Appwrite/Redis/Auth.php:33]: Addressed in f93db69b50. Auth::authenticate() now throws a RedisException when Redis auth returns false, so non-throwing phpredis instances fail fast instead of handing back an unauthenticated connection. I also added a focused unit test covering the false-return path and reran the targeted Redis PHPUnit, PHPStan, and Pint checks.

**Comment by RushikeshGandhmal** [comment]
Hi @lohanidamodar — this PR fixes the Redis NOAUTH issue from #3309 by wiring Redis username/password auth through all Appwrite Redis connection sites, including cache/database, realtime, workers, and queue usage.

The Greptile feedback has been addressed as well: Redis auth now fails fast if phpredis returns false, and the PR includes focused unit/integration coverage for password-only auth, ACL users, nopass users, and the "0" credential edge case.

I would really appreciate a review when you have a chance. Thank you!
